File: tests/test_scene_transport.mjs
import assert from 'node:assert/strict';
import test from 'node:test';
import {fetchSceneResource,sceneJsonParts,sceneLinesParts,iterateSceneLines,writeSceneLines,readSceneLines,SCENE_MANIFEST_FORMAT,SCENE_CHUNK_FORMAT,SCENE_LINES_FORMAT} from '../frontend/scene_transport.js';
import {sceneWithVisibility,sceneVisibilityExport,readIsolation} from '../frontend/scene_visibility.js';

const url='/api/projects/p/files/runs/r/scene.viewer.json';
const records=Array.from({length:5},(_,i)=>({position:[i,.25,-2],source_index:i*13,
  source:i%2?'inferred':'observed',sh_degree:3,sh:Array.from({length:48},(_,j)=>j*.01-i),
  semantic_memberships:['cup','handle'],object_id:`instance-${i}`}));
const header={objects:[{id:'cup',label:'杯子'}],cameras:[{frame_id:'照片.png'}],metadata:{source_gaussian_count:65}};
function fixture(){
  const manifest={format:SCENE_MANIFEST_FORMAT,gaussian_count:5,chunk_size:2,scene:structuredClone(header),
    chunks:[{path:'data/a.json',offset:0,count:2},{path:'data/b.json',offset:2,count:2},{path:'data/c.json',offset:4,count:1}]};
  const pages=new Map([[url,manifest]]);
  for(const chunk of manifest.chunks)pages.set('/api/projects/p/files/runs/r/'+chunk.path,
    {format:SCENE_CHUNK_FORMAT,offset:chunk.offset,count:chunk.count,gaussians:records.slice(chunk.offset,chunk.offset+chunk.count)});
  return {manifest,pages};
}

test('sequential chunks preserve exact SH, semantic identities, row order and header',async()=>{
  const {pages}=fixture();const calls=[];let active=0,peak=0;
  const read=async path=>{calls.push(path);active++;peak=Math.max(peak,active);await Promise.resolve();active--;return pages.get(path);};
  assert.deepEqual(await fetchSceneResource(url,read),{...header,gaussians:records});
  assert.equal(peak,1);assert.equal(calls.length,4);
  assert.deepEqual(calls.slice(1),['a','b','c'].map(n=>'/api/projects/p/files/runs/r/data/'+n+'.json'));
});

test('legacy plain scene needs only one read and remains unchanged',async()=>{
  const scene={...header,gaussians:records};let count=0;
  const result=await fetchSceneResource('/scene.json',async()=>{count++;return scene;});
  assert.equal(result,scene);assert.equal(count,1);
});

test('untrusted remote, encoded, absolute and parent paths rejected before chunk fetch',async()=>{
  for(const path of ['https://evil/a.json','//evil/a.json','/absolute.json','../secret','data/../secret',
                     '%2e%2e/secret','data/a.json?secret','data/a.json#fragment','data\\a.json']){
    const {manifest}=fixture();manifest.chunks[0].path=path;let reads=0;
    await assert.rejects(fetchSceneResource(url,async()=>{reads++;return manifest;}),/路径/);
    assert.equal(reads,1);
  }
});

test('bad count, ordering, duplicate paths and header fail before fetching chunks',async()=>{
  for(const alter of [m=>m.gaussian_count=6,m=>m.gaussian_count=2000001,m=>m.chunks[1].offset=1,
    m=>m.chunks[1].path=m.chunks[0].path,m=>m.chunks[0].count=3,m=>m.scene.gaussians=[],m=>m.chunk_size=25001]){
    const {manifest}=fixture();alter(manifest);let reads=0;
    await assert.rejects(fetchSceneResource(url,async()=>{reads++;return manifest;}));
    assert.equal(reads,1);
  }
});

test('chunk payload count and offset must agree with manifest',async()=>{
  for(const alter of [c=>c.offset=3,c=>c.count=1,c=>c.gaussians.pop(),c=>c.format='other',c=>c.gaussians[0]=null]){
    const {pages}=fixture();const chunk=structuredClone(pages.get('/api/projects/p/files/runs/r/data/a.json'));
    alter(chunk);pages.set('/api/projects/p/files/runs/r/data/a.json',chunk);
    await assert.rejects(fetchSceneResource(url,async path=>pages.get(path)),/清单不一致/);
  }
});

test('unknown manifest format does not become an empty scene',async()=>{
  await assert.rejects(fetchSceneResource(url,async()=>({format:'future',chunks:[]})),/可识别/);
});

test('bounded JSON export parts preserve scene without one giant serialization',()=>{
  const scene={...header,gaussians:records};
  const parts=sceneJsonParts(scene,{chunkSize:2});
  assert.deepEqual(JSON.parse(parts.join('')),scene);assert.ok(parts.length>3);
  assert.deepEqual(JSON.parse(sceneJsonParts({gaussians:[]}).join('')),{gaussians:[]});
  assert.deepEqual(scene.gaussians,records);
});

function byteStream(bytes,width=7){
  let offset=0;
  return new ReadableStream({pull(controller){
    if(offset===bytes.length){controller.close();return;}
    const end=Math.min(bytes.length,offset+width);controller.enqueue(bytes.slice(offset,end));offset=end;
  }});
}
const bytes=text=>new TextEncoder().encode(text);

test('JSONL streamed roundtrip preserves SH colors, source order, unicode and visibility edits',async()=>{
  const base={...header,gaussians:records.map(g=>({...g,semantic_ids:['cup']}))};
  const edited=sceneWithVisibility(base,new Set(['cup']),new Set(['cup']));
  const parts=sceneLinesParts(edited,{chunkSize:2});
  assert.equal(JSON.parse(parts[0]).format,SCENE_LINES_FORMAT);
  assert.ok(parts.every(line=>line.endsWith('\n')));
  // One-byte reads split every multibyte Chinese codepoint and JSON token.
  const loaded=await readSceneLines(byteStream(bytes(parts.join('')),1));
  assert.deepEqual(loaded,edited);
  assert.deepEqual([...readIsolation(loaded,loaded.objects)],['cup']);
  assert.ok(loaded.gaussians.every(g=>g.hidden));
  assert.deepEqual(loaded.gaussians.map(g=>g.source_index),[0,13,26,39,52]);
});

test('JSONL reads browser Blob streams without file.text or one whole-file string',async()=>{
  const scene={...header,gaussians:records};const blob=new Blob(sceneLinesParts(scene,{chunkSize:2}));
  blob.text=()=>{throw new Error('whole-file text must not be read');};
  assert.deepEqual(await readSceneLines(blob),scene);
});

test('JSONL export adaptively reduces chunk count to honor UTF-8 byte limit',async()=>{
  const scene={...header,gaussians:records};
  const single=Math.max(...sceneLinesParts(scene,{chunkSize:1}).map(line=>bytes(line.trimEnd()).length));
  const maximum=single+10;
  const parts=sceneLinesParts(scene,{chunkSize:5,maxLineBytes:maximum});
  assert.ok(parts.length>2);
  assert.ok(parts.every(line=>bytes(line.trimEnd()).length<=maximum));
  assert.deepEqual(await readSceneLines(new Blob(parts),{maxLineBytes:maximum}),scene);
});

test('JSONL truncated file, missing final newline and missing final chunk rejected',async()=>{
  const parts=sceneLinesParts({...header,gaussians:records},{chunkSize:2});
  for(const text of [parts.join('').slice(0,-10),parts.join('').slice(0,-1),parts.slice(0,-1).join('')]){
    await assert.rejects(readSceneLines(byteStream(bytes(text),3)),/截断|缺少/);
  }
});

test('JSONL invalid count, duplicated offset, wrong format and extra chunk rejected',async()=>{
  const parts=sceneLinesParts({...header,gaussians:records},{chunkSize:2});
  const parsed=parts.map(JSON.parse);
  for(const alter of [p=>p[0].gaussian_count=2000001,p=>p[1].offset=1,p=>p[2].offset=0,
                     p=>p[1].count=3,p=>p[1].format='unknown',p=>p.push(p[1])]){
    const values=structuredClone(parsed);alter(values);
    const text=values.map(value=>JSON.stringify(value)+'\n').join('');
    await assert.rejects(readSceneLines(byteStream(bytes(text),17)),/无效|不一致/);
  }
});

test('JSONL oversized line cancels input before unbounded accumulation',async()=>{
  let cancelled=false;
  const stream=new ReadableStream({start(controller){controller.enqueue(bytes(' '.repeat(201)));},
    cancel(){cancelled=true;}});
  await assert.rejects(readSceneLines(stream,{maxLineBytes:200}),/单行超过/);
  assert.ok(cancelled);
});

test('JSONL malformed UTF-8 and blank extra line are rejected',async()=>{
  await assert.rejects(readSceneLines(byteStream(new Uint8Array([0xc3,0x28,10]),1)),/UTF-8/);
  const parts=sceneLinesParts({...header,gaussians:records},{chunkSize:2});
  await assert.rejects(readSceneLines(new Blob([...parts,'\n'])),/有效 JSON/);
});

test('full Bonsai and Teatime counts pass sequential transport without missing source rows',async()=>{
  for(const count of [1065515,1706789]){
    const chunkSize=25000,chunks=[];let calls=0,progress=0;
    for(let offset=0;offset<count;offset+=chunkSize)chunks.push({path:`chunks/${offset}.json`,offset,count:Math.min(chunkSize,count-offset)});
    const manifest={format:SCENE_MANIFEST_FORMAT,gaussian_count:count,chunk_size:chunkSize,scene:{objects:[]},chunks};
    const sh=Array.from({length:48},(_,i)=>i/100);
    const loaded=await fetchSceneResource('/scene.viewer.json',async url=>{
      calls++;if(url==='/scene.viewer.json')return manifest;
      const offset=Number(url.match(/(\d+)\.json$/)[1]),entry=chunks[offset/chunkSize];
      return {format:SCENE_CHUNK_FORMAT,...entry,gaussians:Array.from({length:entry.count},(_,i)=>({source_index:offset+i,sh_degree:3,sh}))};
    },{onProgress:value=>{assert.ok(value.loaded>=progress);progress=value.loaded;}});
    assert.equal(loaded.gaussians.length,count);assert.equal(progress,count);assert.equal(calls,chunks.length+1);
    for(let i=0;i<count;i++)assert.equal(loaded.gaussians[i].source_index,i);
    assert.equal(loaded.gaussians[count-1].sh,sh);
  }
});

test('plain scenes and invalid caller capacities cannot bypass the full-scene safety guard',async()=>{
  await assert.rejects(fetchSceneResource('/scene.json',async()=>({gaussians:new Array(2000001)})),/超过当前完整显示上限/);
  await assert.rejects(fetchSceneResource('/scene.json',async()=>({gaussians:[]}),{maxGaussians:2000001}),/上限无效/);
});

test('streaming export applies visibility per chunk and honors writer backpressure',async()=>{
  const original={...header,gaussians:records.map(g=>({...g,semantic_ids:['cup']}))};
  const prepared=sceneVisibilityExport(original,new Set(['cup']),new Set(['cup']));
  assert.equal(prepared.scene.gaussians,original.gaussians);
  let mapped=0,pending=false,closed=false;const written=[];
  await writeSceneLines(prepared.scene,{async write(line){assert.equal(pending,false);pending=true;const before=mapped;await Promise.resolve();assert.equal(mapped,before);written.push(line);pending=false;},async close(){closed=true;}},
    {chunkSize:2,mapGaussian:g=>{mapped++;return prepared.mapGaussian(g);}});
  assert.equal(mapped,records.length);assert.ok(closed);
  const restored=await readSceneLines(new Blob(written));
  assert.deepEqual(restored,sceneWithVisibility(original,new Set(['cup']),new Set(['cup'])));
  assert.ok(original.gaussians.every(g=>g.hidden===undefined));
  let lazyCount=0;const iterator=iterateSceneLines(original,{chunkSize:2,mapGaussian:g=>{lazyCount++;return g;}});
  iterator.next();assert.equal(lazyCount,0);iterator.next();assert.equal(lazyCount,2);iterator.return();assert.equal(lazyCount,2);
});

test('failed streaming export aborts without creating subsequent chunks or closing successfully',async()=>{
  let writes=0,mapped=0,aborted=false,closed=false;
  await assert.rejects(writeSceneLines({...header,gaussians:records},{write(){if(++writes===2)throw new Error('disk full');},close(){closed=true;},abort(){aborted=true;}},
    {chunkSize:2,mapGaussian:g=>{mapped++;return g;}}),/disk full/);
  assert.equal(mapped,2);assert.equal(writes,2);assert.ok(aborted);assert.equal(closed,false);
});

File: tests/test_scene_transport.py
"""Exact, bounded JSON transport and PLY import budgets; no GPU execution."""
import copy
import json
from pathlib import Path
import sys

import pytest
from fastapi.testclient import TestClient

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))
from backend.scene_transport import CHUNK_FORMAT, MANIFEST_FORMAT, LINES_FORMAT, read_scene_lines, write_viewer_scene


def scene(count=5):
    return {'gaussians': [{'position': [i, .5, -2], 'color': [.1, .2, .3], 'scale': [.1]*3,
                           'rotation': [1, 0, 0, 0], 'opacity': .7, 'source_index': i*7,
                           'source': 'inferred' if i%2 else 'observed', 'sh_degree': 3,
                           'sh': [j*.00001-i*.01 for j in range(48)],
                           'semantic_memberships': ['cup', 'handle'] if i%2 else [],
                           'object_id': f'instance-{i}'} for i in range(count)],
            'cameras': [{'frame_id': '真实帧.png'}], 'objects': [{'id': 'cup', 'label': '杯子'}],
            'metadata': {'source_gaussian_count': count*7, 'preview_sampled': True}}


def test_chunks_preserve_all_values_global_order_and_header(tmp_path):
    original = scene()
    canonical = tmp_path/'scene.json'
    canonical.write_text(json.dumps(original))
    before = canonical.read_bytes()
    result = write_viewer_scene(original, canonical, threshold=2, chunk_size=2)
    manifest = json.loads(result.read_text())
    assert manifest['format'] == MANIFEST_FORMAT
    assert manifest['gaussian_count'] == 5
    assert [row['offset'] for row in manifest['chunks']] == [0, 2, 4]
    assert [row['count'] for row in manifest['chunks']] == [2, 2, 1]
    actual = []
    for row in manifest['chunks']:
        path = tmp_path/row['path']
        assert path.is_relative_to(tmp_path)
        chunk = json.loads(path.read_text())
        assert chunk['format'] == CHUNK_FORMAT
        assert chunk['offset'] == len(actual)
        assert chunk['count'] == len(chunk['gaussians'])
        actual.extend(chunk['gaussians'])
    assert {**manifest['scene'], 'gaussians': actual} == original
    assert canonical.read_bytes() == before
    assert [g['source_index'] for g in actual] == [0, 7, 14, 21, 28]


def test_small_scene_keeps_legacy_path_without_extra_writes(tmp_path):
    canonical = tmp_path/'scene.json'
    canonical.write_text('{}')
    assert write_viewer_scene(scene(), canonical) == canonical
    assert list(tmp_path.iterdir()) == [canonical]


def test_default_threshold_and_25000_record_bound(tmp_path):
    original = {'gaussians': [{'source_index': i} for i in range(50001)]}
    path = write_viewer_scene(original, tmp_path/'scene.json')
    manifest = json.loads(path.read_text())
    assert [row['count'] for row in manifest['chunks']] == [25000, 25000, 1]


def test_republication_does_not_overwrite_chunks_already_in_use(tmp_path):
    original = scene()
    first = write_viewer_scene(original, tmp_path/'scene.json', threshold=1, chunk_size=2)
    prior = json.loads(first.read_text())
    snapshots = {(tmp_path/row['path']): (tmp_path/row['path']).read_bytes() for row in prior['chunks']}
    changed = scene();changed['gaussians'][0]['source_index'] = 100
    second = write_viewer_scene(changed, tmp_path/'scene.json', threshold=1, chunk_size=2)
    current = json.loads(second.read_text())
    assert {r['path'] for r in prior['chunks']}.isdisjoint(r['path'] for r in current['chunks'])
    assert all(path.read_bytes() == value for path, value in snapshots.items())


def test_failed_serialization_preserves_previous_manifest_and_cleans_own_chunks(tmp_path):
    original = scene()
    path = write_viewer_scene(original, tmp_path/'scene.json', threshold=1, chunk_size=2)
    before, entries = path.read_bytes(), set(tmp_path.iterdir())
    original['gaussians'][-1]['sh'][0] = float('nan')
    with pytest.raises(ValueError):
        write_viewer_scene(original, tmp_path/'scene.json', threshold=1, chunk_size=2)
    assert path.read_bytes() == before
    assert set(tmp_path.iterdir()) == entries


@pytest.mark.parametrize('size', [0, -1, 25001, True])
def test_chunk_limit_cannot_be_bypassed(tmp_path, size):
    with pytest.raises(ValueError):
        write_viewer_scene(scene(), tmp_path/'scene.json', chunk_size=size)


@pytest.fixture
def api_client(tmp_path, monkeypatch):
    import backend.app as app_module
    monkeypatch.setattr(app_module, 'DATA', tmp_path)
    return app_module, TestClient(app_module.app)


@pytest.mark.parametrize('mode,expected', [('full',None), ('fast',100000), ('balanced',250000), ('fine',500000)])
def test_ply_import_passes_selected_budget_and_exposes_chunk_manifest(api_client, monkeypatch, mode, expected):
    module, client = api_client
    calls = []
    def read(path, max_points):
        calls.append(max_points)
        return scene(3)
    monkeypatch.setattr(module, 'read_ply', read)
    monkeypatch.setattr(module, 'write_viewer_scene', lambda value, path: write_viewer_scene(value, path, threshold=1, chunk_size=2))
    result = client.post('/api/import', files={'file': ('model.ply', b'ply mock')}, data={'mode': mode})
    assert result.status_code == 200, result.text
    body = result.json()
    assert calls == [expected]
    assert body['scene_url'].endswith('/scene.json')
    assert body['viewer_scene_url'].endswith('/scene.viewer.json')
    canonical = client.get(body['scene_url']).json()
    assert canonical['metadata']['preview_point_budget'] == expected
    assert canonical['metadata']['preview_mode'] == mode
    manifest = client.get(body['viewer_scene_url']).json()
    assert manifest['gaussian_count'] == 3
    assert manifest['scene']['metadata']['source_gaussian_count'] == 21


def test_unknown_ply_mode_rejected_before_reader_called(api_client, monkeypatch):
    module, client = api_client
    def forbidden(*args):
        raise AssertionError('invalid mode reached PLY reader')
    monkeypatch.setattr(module, 'read_ply', forbidden)
    result = client.post('/api/import', files={'file': ('model.ply', b'ply mock')}, data={'mode':'ultra'})
    assert result.status_code == 422


def test_default_ply_mode_is_full_and_small_json_stays_backward_compatible(api_client, monkeypatch):
    module, client = api_client
    calls = []
    monkeypatch.setattr(module, 'read_ply', lambda path, maximum: calls.append(maximum) or scene(1))
    result = client.post('/api/import', files={'file': ('model.ply', b'ply mock')})
    assert calls == [None]
    assert result.status_code == 200
    result = client.post('/api/import', files={'file': ('model.json', json.dumps(scene(1)), 'application/json')})
    body = result.json()
    assert body['scene_url'] == body['viewer_scene_url']
    assert client.get(body['scene_url']).json()['gaussians'] == scene(1)['gaussians']


def test_full_resource_limit_rejects_without_publishing(api_client, monkeypatch, tmp_path):
    module, client = api_client
    import backend.scene_limits as limits
    monkeypatch.setattr(limits, 'MAX_SCENE_GAUSSIANS', 2)
    monkeypatch.setattr(module, 'read_ply', lambda *args: scene(3))
    response = client.post('/api/import', files={'file': ('full.ply', b'ply mock')})
    assert response.status_code == 422
    assert '不会自动抽样' in response.text
    assert not list(tmp_path.iterdir())


def test_streamed_upload_limit_removes_partial_project(api_client, monkeypatch, tmp_path):
    module, client = api_client
    import backend.scene_limits as limits
    monkeypatch.setattr(limits, 'MAX_PLY_IMPORT_BYTES', 5)
    response = client.post('/api/import', files={'file': ('too-large.ply', b'123456')})
    assert response.status_code == 413
    assert not list(tmp_path.iterdir())


def test_limits_accept_fine_scene_and_reject_over_capacity_without_sampling():
    from backend.scene_limits import validate_gaussian_count, scene_point_budget
    validate_gaussian_count(2_162_260)
    validate_gaussian_count(3_000_000)
    assert scene_point_budget() is None
    with pytest.raises(ValueError, match='不会自动抽样'):
        validate_gaussian_count(3_000_001)


def test_all_training_modes_default_to_full_display():
    from backend.app import Config
    from backend.scene_limits import scene_point_budget
    for mode in ('fast','balanced','fine'):
        assert scene_point_budget(Config(mode=mode).viewer_quality) is None


def lines_bytes(original):
    header={key:value for key,value in original.items() if key!='gaussians'}
    records=[{'format':LINES_FORMAT,'gaussian_count':len(original['gaussians']),'scene':header}]
    for offset in range(0,len(original['gaussians']),2):
        values=original['gaussians'][offset:offset+2]
        records.append({'format':CHUNK_FORMAT,'offset':offset,'count':len(values),'gaussians':values})
    return ('\n'.join(json.dumps(value,ensure_ascii=False) for value in records)+'\n').encode('utf-8')


def test_jsonl_import_preserves_sh_global_rows_edits_and_metadata(api_client):
    module,client=api_client
    original=scene(5);original['gaussians'][2].update(hidden=True,priority=True)
    response=client.post('/api/import',files={'file':('complete.splat.jsonl',lines_bytes(original))})
    assert response.status_code==200,response.text
    result=client.get(response.json()['scene_url']).json()
    assert result['gaussians']==original['gaussians']
    assert result['objects']==original['objects']
    assert result['cameras']==original['cameras']
    assert all(result['metadata'][key]==value for key,value in original['metadata'].items())


def test_jsonl_larger_than_ply_and_json_cap_still_imports(api_client,monkeypatch):
    import backend.scene_limits as limits
    module,client=api_client
    monkeypatch.setattr(limits,'MAX_PLY_IMPORT_BYTES',5)
    monkeypatch.setattr(limits,'MAX_JSON_IMPORT_BYTES',5)
    response=client.post('/api/import',files={'file':('full.jsonl',lines_bytes(scene(3)))})
    assert response.status_code==200,response.text


def test_format_limits_cover_real_full_bonsai_export_without_large_fixture():
    from backend.scene_limits import import_file_limit,MAX_SCENE_LINE_BYTES
    assert import_file_limit('.ply')==1024**3
    assert import_file_limit('.json')==256*1024**2
    assert import_file_limit('.jsonl')==4*1024**3
    assert import_file_limit('.jsonl')>1_372_757_800
    assert MAX_SCENE_LINE_BYTES==64*1024**2


@pytest.mark.parametrize('alter',[
    lambda value:value[:-1],
    lambda value:value.replace(b'"offset": 2',b'"offset": 1'),
    lambda value:value.replace(b'"count": 2',b'"count": 3',1),
    lambda value:value.replace(b'"gaussian_count": 5',b'"gaussian_count": 2000001'),
    lambda value:value+b'{}\n',
    lambda value:value.replace('杯子'.encode(),b'\xff'),
    lambda value:value.split(b'\n',1)[0]+b'\n',
])
def test_jsonl_rejects_truncated_misaligned_oversized_or_invalid_input(tmp_path,alter):
    path=tmp_path/'invalid.jsonl';path.write_bytes(alter(lines_bytes(scene(5))))
    with pytest.raises(ValueError):read_scene_lines(path)


def test_jsonl_byte_line_cap_is_enforced_before_decode(tmp_path):
    path=tmp_path/'long.jsonl';path.write_bytes(lines_bytes(scene()))
    with pytest.raises(ValueError,match='单行'):
        read_scene_lines(path,max_line_bytes=32)
